docs/functionaldatatypes/nonemptylist.md

```diff
@@ -0,0 +1,195 @@
+
+For example, we might be handling a collection of errors.
+
+If an error has occurred at all then we know the collection is not empty. However, the Scala compiler does not know that and so we are forced to handle the case where the collection is empty even though we know that cannot occur.
+
+Or perhaps we are implementing a batched API and need to handle a collection of requests. We know the collection is not empty or we would not have been called at all, but now we have to handle the case of an empty collection.
+
+Sometimes there are sensible defaults in these cases, but other times we are forced to resort to potentially unsafe operators like `head` or to throw exceptions in branches of logic we know should never occur. This wastes developer time, makes code harder to understand, and encourages bad coding practices.
+
+To avoid these problems we want a data type that is specific enough to describe at the type level what we know to be true, in this case that a collection is not empty.
+
+We can try to represent this using the `::` case of `List` from the Scala standard library but unfortunately `::` itself has extremely poor ergonomics. It is hard to construct `::` values and it is easy for operators to forget the more specific type.
+
+This is where `NonEmptyList` comes in.
+
+A `NonEmptyList` looks like this:
+
+```scala mdoc
+sealed trait NonEmptyList[+A]
+
+object NonEmptyList {
+  case class Single[A](head: A) extends NonEmptyList[A]
+  case class Cons[A](head: A, tail: NonEmptyList[A]) extends NonEmptyList[A]
+}
+```
+
+We can see that `NonEmptyList` looks very similar to `List` except that there is no `Nil` case. A `NonEmptyList` is either a `Single` with a single value of type `A` or a `Cons` with a value of type `A` and a `NonEmptyList` of further values of type `A`.
+
+Generally you should just be able to use higher level operators on `NonEmptyList` but if you ever want to implement your own operators on `NonEmptyList` you can do so by pattern matching on a `NonEmptyList` just like you would on a `List`.
+
+## Constructing NonEmptyList Values
+
+The easiest way to construct a `NonEmptyList` is with the `apply` operator.
+
+```scala mdoc:nest
+trait NonEmptyList[+A]
+
+object NonEmptyList {
+  def apply[A](a: A, as: A*): NonEmptyList[A] =
+    ???
+}
+```
+
+If we already have another collection type we can convert it into a `NonEmptyList` using the `fromIterableOption` operator.
+
+```scala mdoc:nest
+object NonEmptyList {
+  def fromIterableOption[A](as: Iterable[A]): Option[NonEmptyList[A]] =
+    ???
+}
+```
+
+Notice that `fromIterableOption` returns an `Option[NonEmptyList[A]]` because the original collection might be empty and so we might not be able to construct a `NonEmptyList` from it.
+
+If we know the collection is not empty we can use the `fromIterable` operator and provide the first element of the collection to "prove" that it is not empty.
```

Comment:
pedantic: I find double quotes around prove distracting, you may drop them. Alternatively, you could say "to satisfy the NonEmptyList type requirement".

Reply:
Changed to "to establish that it is not empty".